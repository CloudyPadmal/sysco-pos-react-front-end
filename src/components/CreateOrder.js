import React, { Component } from 'react';
import { withRouter, Redirect } from 'react-router-dom'
import ListItemInEditOrder from './ListItemInEditOrder';

/**
 * @abstract Creates a new order
 * @description Creates a new order with multiple items added to it
 * @param ISLOGGEDIN
 * @param CURRENTORDERID
 * @param ITEMSLIST
 * @param DELETE_THIS_ORDER
 * @param CREATE_NEW_ORDER_FOR_THIS_USER
 * @param ADD_ITEMS_TO_THIS_ORDER
 */
class CreateOrder extends Component {

    constructor(props) {
        super(props);
        this.state = {
            TOTAL: 0
        };
    }

    // As the create order view mounted, generate a new order for the user
    // and in the meantime fetch the complete item list
    componentDidMount() {
        this.props.CREATE_NEW_ORDER_FOR_THIS_USER();
    }

    /**************************************************************************
     * Once all the items are set, this method will trigger the order state 
     * updating function in parent component to trigger a series of axios 
     * requests adding the items selected to the order and update the global 
     * item count. Before all that, this will filter the items for zero item 
     * counts. Once the parent function is triggered, redirects user to order
     * list after waiting for a time out depending on the number of items being
     * added to show user the completed order rather than a half filled order
     * as it takes some time to complete the axios promises.
     *************************************************************************/
    ADD_THESE_ITEMS_TO_THIS_ORDER = () => {
        this.props.ADD_ITEMS_TO_THIS_ORDER();
        setTimeout(() => {
            this.props.history.push('/my_orders')
        }, (50 * this.props.ITEMQUANTITY.length));
    }

    /**************************************************************************
     * If user chose not to create this order, this local function will trigger
     * parent component function to delete the current order being created. 
     * Once done, user will be redirected to order list.
     *************************************************************************/
    CANCEL_THE_ORDER = (ID) => {
        this.props.DELETE_THIS_ORDER(ID);
        this.props.CLEAR_ORDER_ADDING_PROCESS();
        this.props.history.push('/my_orders');
    }

    render() {
        if (!this.props.ISLOGGEDIN) {
            return (
                <Redirect to="/login" />
            );
        } else {
            return (
                <div>
                    <div className="card" style={{ margin: '25px', paddingBottom: '50px' }}>
                        <div className="card-body">
                            <h5 className="card-title"
                                style={{ margin: '0.5em 1em' }}>Item List</h5>
                            {this.props.ITEMSLIST.map((item) => (
                                <ListItemInEditOrder
                                    key={item._id}
                                    ITEM={item}
                                    NAME={item.productID}
                                    ITEMQUANTITY={this.props.ITEMQUANTITY[item.productID] === undefined
                                        ? 0 : this.props.ITEMQUANTITY[item.productID]}
                                    ADD_THIS_ITEM_TO_ITEMQUANTITY={this.props.ADD_THIS_ITEM_TO_ITEMQUANTITY}
                                    DELETE_THIS_ITEM={this.props.DELETE_THIS_ITEM}
                                    INDECCREMENT_ITEM_COUNT={this.props.INDECCREMENT_ITEM_COUNT} />
                            ))}
                        </div>
                    </div>
                    <nav className="navbar navbar-light bg-light"
                        style={{
                            paddingRight: '25px',
                            overflow: 'hidden',
                            position: 'fixed',
                            zIndex: 99,
                            bottom: '0',
                            width: '100%'
                        }}>
                        <div className="card-body" style={{ paddingTop: '0px' }}>
                            <div className="row">
                                <div className="col-12 d-flex justify-content-end">
                                    <button
                                        onClick={this.ADD_THESE_ITEMS_TO_THIS_ORDER}
                                        className="btn btn-success"
                                        style={{ marginRight: '10px' }}>Add to Cart</button>
                                    <button
                                        onClick={this.CANCEL_THE_ORDER.bind(
                                            this, this.props.CURRENTORDERID
                                        )}
                                        className="btn btn-danger">Cancel</button>
                                </div>
                            </div>
                        </div>
                    </nav>
                </div>
            );
        }
    }
}

export default withRouter(CreateOrder);
