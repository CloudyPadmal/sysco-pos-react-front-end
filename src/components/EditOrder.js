import React, { Component } from 'react';
import { withRouter, Redirect } from 'react-router-dom'
import ListItemInEditOrder from './ListItemInEditOrder';

class EditOrder extends Component {

    constructor(props) {
        super(props);
        this.state = {
            UPDATEDITEMS: {},
            ITEMQUANTITYINORDER: {}
        };
    }

    // Once the component is mounted, filter out the productID and quantity of
    // the items in order and append it to local state
    componentDidMount() {
        let { items } = this.props.CURRENTORDER;
        let ITEMQUANTITYINORDER = {};
        for (var i in items) {
            ITEMQUANTITYINORDER[items[i].productID] = items[i].quantity;
        }
        this.setState({ ITEMQUANTITYINORDER });
    }

    /**************************************************************************
     * 
     *************************************************************************/
    UPDATE_THIS_ORDER = () => {
        let { UPDATEDITEMS } = this.state;
        let REFINEDITEMS = {};
        for (var ID in UPDATEDITEMS) {
            if (parseInt(UPDATEDITEMS[ID]) !== 0) {
                REFINEDITEMS[ID] = UPDATEDITEMS[ID];
            }
        }
        // TODO: Calculate differences
        this.props.UPDATE_ITEMS_IN_THIS_ORDER(REFINEDITEMS);
        setTimeout(() => {
            this.props.history.push('/my_orders')
        }, (50 * REFINEDITEMS.length));
    }

    /**************************************************************************
     * Do not make any modifications to the current order or the order list,
     * just redirect user back to order list
     *************************************************************************/
    CANCEL_EDITING_THE_ORDER = () => {
        this.props.history.push('/my_orders');
    }

    /**************************************************************************
     * Local function to add a single item to a temporary list of items to be 
     * added in to the order being updated. This will take care of duplicate 
     * entries but not zero entries. We'll need to refine that when updating
     * @param ID productID of the item being added
     * @param VALUE quantity of the item added
     *************************************************************************/
    ADD_THIS_ITEM_TO_ORDER = (ID, VALUE) => {
        try {
            delete this.state.UPDATEDITEMS.id;
        } catch (e) { }
        let tempAddedItems = this.state.UPDATEDITEMS;
        tempAddedItems[ID] = VALUE;
        this.setState({
            addedItems: tempAddedItems
        });
    }

    render() {
        if (this.props.CURRENTORDERID === "") {
            return (
                <Redirect to="/my_orders" />
            );
        }
        return (
            <div className="card" style={{ margin: '25px' }}>
                <div className="card-body">
                    <h5 className="card-title" style={{ margin: '0.5em 1em' }}>Editing Order {this.props.CURRENTORDERID}</h5>
                    {this.props.ITEMSLIST.map((item) => (
                        <ListItemInEditOrder
                            key={item._id}
                            ITEM={item}
                            keyID={item.productID}
                            ADD_THIS_ITEM_TO_ORDER={this.ADD_THIS_ITEM_TO_ORDER}
                            ITEMQTYINORDER={this.state.ITEMQUANTITYINORDER[item.productID]} />
                    ))}
                </div>
                <div className="card-body button-section" style={{ paddingTop: '0px' }}>
                    <div className="row">
                        <div className="col-12 d-flex justify-content-end">
                            <button onClick={this.UPDATE_THIS_ORDER} className="btn btn-primary" style={{ marginRight: '10px' }}>Update Order</button>
                            <button onClick={this.CANCEL_EDITING_THE_ORDER} className="btn btn-danger">Cancel</button>
                        </div>
                    </div>
                </div>
            </div>
        );
    }
}

export default withRouter(EditOrder);
