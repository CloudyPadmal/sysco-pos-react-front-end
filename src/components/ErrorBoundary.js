import React from 'react';
import { Link } from 'react-router-dom'

class ErrorBoundary extends React.Component {

    constructor(props) {
        super(props);
        this.state = { hasError: false };
    }

    static getDerivedStateFromError(error) {
        // Update state so the next render will show the fallback UI.
        return { hasError: true };
    }

    render() {
        if (this.state.hasError) {
            return <h1 className="d-flex justify-content-center"
                style={{ height: '100%', margin: '30% 0 0 0' }}>
                <span className="badge badge-dark" style={{ padding: '50px' }}>
                    Cannot load the website ... <br /><br />Want to go back to
                    <br /><br /><Link to="/home">Home Page</Link>?</span>
            </h1>;
        }
        return this.props.children;
    }
}

export default ErrorBoundary;