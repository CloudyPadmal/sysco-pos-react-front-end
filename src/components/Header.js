import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

export function ConditionalHeader(P) {
    if (P.isLoggedIn) {
        return (
            <div className="collapse navbar-collapse">
                <Link className="navbar-brand"
                    style={{ paddingTop: '0px' }}
                    to="/home">Sysco PoS System</Link>
                <div className="navbar-nav mr-auto mt-2 mt-lg-0">
                    <Link className="nav-item nav-link"
                        to="/my_orders">
                        <i className="fas fa-list-ul"></i> My Orders</Link>
                    <Link className="nav-item nav-link"
                        to="/create_order">
                        <i className="fas fa-plus"></i> Create Order</Link>
                </div>
                <div className="navbar-nav my-2 my-lg-0">
                    <Link className="nav-item nav-link"
                        to="/logout">
                        <i className="fas fa-sign-out-alt"></i> Logout</Link>
                </div>
            </div>
        );
    } else {
        return (
            <div className="collapse navbar-collapse">
                <Link className="navbar-brand"
                    style={{ paddingTop: '0px' }}
                    to="/">Sysco PoS System</Link>
                <div className="navbar-nav ml-auto">
                    <Link className="nav-item nav-link"
                        to="/login">
                        <i className="fas fa-sign-in-alt"></i> Login</Link>
                    <Link className="nav-item nav-link"
                        to="/register">Register</Link>
                </div>
            </div>
        );
    }
}

/******************************************************************************
 * Navigation bar at the top of the page. This will change state depending on
 * the log state of user. The final render will be conditioned on that and will
 * be rendered using `ConditionalHeader` component in this same component class
 *****************************************************************************/
export class Header extends Component {
    render() {
        return (
            <header>
                <nav className="navbar navbar-expand-lg navbar-light bg-light">
                    <ConditionalHeader isLoggedIn={this.props.isLoggedIn} />
                </nav>
            </header>
        );
    }
}

Header.propTypes = {
    isLoggedIn: PropTypes.bool.isRequired
};

const mapStateToProps = state => ({
    isLoggedIn: state.uac.isLoggedIn
});

export default connect(mapStateToProps, null)(Header)