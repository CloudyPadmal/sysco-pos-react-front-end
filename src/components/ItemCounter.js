import React, { Component } from 'react';

/**
 * Custom component to increment and decrement value within a minimum and
 * maximum range. A local count will be maintained to help with validation
 * @param NAME Identifier for this counter
 * @param INITIAL Initial value to display
 * @param MIN Minimum allowable value
 * @param MAX Maximum allowable value
 * @param ADD_THIS_ITEM Function responsible for handling the item counts
 */
class ItemCounter extends Component {

    state = {
        count: 0
    }

    INCREMENT_COUNT = (e) => {
        e.preventDefault();
        if (this.state.count !== this.props.MAX) {
            let newCount = this.state.count + 1;
            this.setState({
                count: newCount
            });
            this.props.ADD_THIS_ITEM(this.props.NAME, newCount);
        }
    }

    DECREMENT_COUNT = (e) => {
        e.preventDefault();
        if (this.state.count !== this.props.MIN) {
            let newCount = this.state.count - 1;
            this.setState({
                count: newCount
            });
            this.props.ADD_THIS_ITEM(this.props.NAME, newCount);
        }
    }

    render() {
        return (
            <React.Fragment>
                <div className="btn-group" role="group" aria-label="First group">
                    <button type="button" className="btn btn-secondary" onClick={this.INCREMENT_COUNT}>&#43;</button>
                    <div style={{ width: '3em', textAlign: 'center' }}>
                        <span className="input-group-text" style={{
                            display: 'inline-block', borderRadius: '0px', width: '100%'
                        }}>{this.state.count}</span></div>
                    <button type="button" className="btn btn-secondary" onClick={this.DECREMENT_COUNT}>&#8722;</button>
                </div>
            </React.Fragment>
        );
    }
}

export default ItemCounter;
