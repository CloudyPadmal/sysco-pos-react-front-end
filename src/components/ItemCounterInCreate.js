import React, { Component } from 'react';

/**
 * Custom component to increment and decrement value within a minimum and
 * maximum range. A local count will be maintained to help with validation
 * @param NAME Identifier for this counter
 * @param INITIAL Initial value to display
 * @param MIN Minimum allowable value
 * @param MAX Maximum allowable value
 * @param ADD_THIS_ITEM Function responsible for handling the item counts
 */
class ItemCounterInCreate extends Component {

    constructor(props) {
        super(props);
        this.state = {
            count: props.INITIAL || 0
        }
    }

    /**************************************************************************
     * This will add one to the local count until it reaches the value defined
     * by MAX value and triggers the ADD_THIS_ITEM function set by parent
     *************************************************************************/
    INCREMENT_COUNT = (e) => {
        e.preventDefault();
        if (this.state.count !== this.props.MAX) {
            let newCount = this.state.count + 1;
            this.setState({
                count: newCount
            });
            this.props.ADD_THIS_ITEM(this.props.NAME, newCount);
        }
    }

    /**************************************************************************
     * This will substract one from the local count until it reaches the value
     * defined by MIN value and triggers the ADD_THIS_ITEM function
     *************************************************************************/
    DECREMENT_COUNT = (e) => {
        e.preventDefault();
        if (this.state.count !== this.props.MIN) {
            let newCount = this.state.count - 1;
            this.setState({
                count: newCount
            });
            this.props.ADD_THIS_ITEM(this.props.NAME, newCount);
        }
    }

    render() {
        return (
            <div className="btn-toolbar" role="toolbar" aria-label="Toolbar with button groups">
                <div className="btn-group" role="group" aria-label="Item-Group" style={{ marginRight: '10px' }}>
                    <button type="button" className="btn btn-secondary" onClick={this.INCREMENT_COUNT}>&#43;</button>
                    <div style={{ width: '3em', textAlign: 'center' }}>
                        <span className="input-group-text" style={{
                            display: 'inline-block', borderRadius: '0px', width: '100%', height: '100%'
                        }}>{this.state.count}</span></div>
                    <button type="button" className="btn btn-secondary" onClick={this.DECREMENT_COUNT}>&#8722;</button>
                </div>
            </div>
        );
    }
}

export default ItemCounterInCreate;
