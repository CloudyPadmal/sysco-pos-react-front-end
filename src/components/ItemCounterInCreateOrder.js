import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import {
    dispatch_AddThisItemToQuantity,
    dispatch_DeleteThisItemFromQuantity,
    dispatch_IndecrementItemFromQuantity
} from '../actions/ordercontrolactions';

function RenderDeleteButton(props) {
    if (props.quantity !== 0) {
        return (
            <div className="btn-group"
                role="group" aria-label="Delete-Group" style={{ marginLeft: '10px' }}>
                <button type="button"
                    className="btn btn-danger"
                    onClick={props.DELETE_THIS_ITEM}><i className="fas fa-trash"></i></button>
            </div>
        );
    } else {
        return (
            <React.Fragment></React.Fragment>
        );
    }
}

/**
 * Custom component to increment and decrement value within a minimum and
 * maximum range. A local count will be maintained to help with validation
 * @param name Identifier for this counter
 * @param min Minimum allowable value
 * @param max Maximum allowable value
 */
class ItemCounterInCreateOrder extends Component {

    /**************************************************************************
     * Triggers the parent component method to clear the count for this item
     *************************************************************************/
    deleteThisItem = (e) => {
        e.preventDefault();
        let P = this.props.itemsList.find(
            I => I.productID === this.props.name).price;
        this.props.dispatch_DeleteThisItemFromQuantity(this.props.name, P);
    }

    /**************************************************************************
     * This will increment the item count by one if it is an existing item and
     * will request to add a new item to the quantity list if it is not and do
     * update its count as usual
     *************************************************************************/
    incrementCount = (e) => {
        e.preventDefault();
        if (this.props.quantity !== this.props.max) {
            if (this.props.quantity === 0) {
                this.props.dispatch_AddThisItemToQuantity(this.props.name);
            }
            let P = this.props.itemsList.find(
                I => I.productID === this.props.name).price;
            this.props.dispatch_IndecrementItemFromQuantity(this.props.name, true, P);
        }
    }

    /**************************************************************************
     * This method will decrement the count of item quantities in parent state
     *************************************************************************/
    decrementCount = (e) => {
        e.preventDefault();
        if (this.props.quantity > this.props.min) {
            let P = this.props.itemsList.find(
                I => I.productID === this.props.name).price;
            this.props.dispatch_IndecrementItemFromQuantity(this.props.name, false, P);
        } else if (this.props.quantity === this.props.min) {
            this.deleteThisItem(e);
        }
    }

    render() {
        return (
            <React.Fragment>
                <div className="btn-toolbar" role="toolbar" aria-label="Toolbar with button groups">
                    <div className="btn-group" role="group" aria-label="Item-Group">
                        <button type="button" className="btn btn-secondary"
                            onClick={this.incrementCount}>&#43;</button>
                        <div style={{ width: '3em', textAlign: 'center' }}>
                            <span className="input-group-text" style={{
                                display: 'inline-block', borderRadius: '0px', width: '100%', height: '100%'
                            }}>{this.props.quantity}</span></div>
                        <button type="button" className="btn btn-secondary"
                            onClick={this.decrementCount}>&#8722;</button>
                    </div>
                    <RenderDeleteButton
                        quantity={this.props.quantity}
                        DELETE_THIS_ITEM={this.deleteThisItem} />
                </div>
            </React.Fragment>
        );
    }
}

ItemCounterInCreateOrder.propTypes = {
    dispatch_AddThisItemToQuantity: PropTypes.func.isRequired,
    dispatch_IndecrementItemFromQuantity: PropTypes.func.isRequired,
    dispatch_DeleteThisItemFromQuantity: PropTypes.func.isRequired
};

const mapStateToProps = (state) => ({
    isLoggedIn: state.uac.isLoggedIn,
    currentOrderID: state.ord.currentOrderID,
    passKey: state.uac.passKey,
    itemQuantity: state.ord.itemQuantity,
    itemsList: state.ord.itemsList,
    total: state.ord.total
});

export default connect(mapStateToProps, {
    dispatch_AddThisItemToQuantity,
    dispatch_IndecrementItemFromQuantity,
    dispatch_DeleteThisItemFromQuantity
})(ItemCounterInCreateOrder);
