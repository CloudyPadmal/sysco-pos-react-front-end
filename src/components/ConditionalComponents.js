import React from 'react';
import { Redirect } from 'react-router-dom';

import EditOrder from './EditOrder';

export function DecidedLandingPage(prop) {
    if (prop.ISLOGGEDIN) {
        return (
            <Redirect to='/my_orders' />
        );
    } else {
        return (
            <Redirect to='/login' />
        );
    }
}

export function GoToEditOrder(prop) {
    if (prop.currentOrderInContext) {
        return (
            <EditOrder
                usertoken={prop.usertoken}
                viewingOrder={prop.viewingOrder}
                fetchAllItemsList={prop.fetchAllItemsList}
                currentOrderInContext={prop.currentOrderInContext}
                allItemsList={prop.allItemsList} />
        );
    } else {
        return (
            <Redirect to='/my_orders' />
        );
    }
}