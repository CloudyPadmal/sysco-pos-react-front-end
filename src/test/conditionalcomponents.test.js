import React from 'react';
import { shallow } from './enzyme';

import { DecidedLandingPage } from '../components/ConditionalComponents';

describe('Deciding Landing Page', () => {

  it('renders the order list when user is logged in', () => {
    const wrapper = shallow(<DecidedLandingPage ISLOGGEDIN={true} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders the login page when user is not logged in', () => {
    const wrapper = shallow(<DecidedLandingPage ISLOGGEDIN={false} />);
    expect(wrapper).toMatchSnapshot();
  });

});