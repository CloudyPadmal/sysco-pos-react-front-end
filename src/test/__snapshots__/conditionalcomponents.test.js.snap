// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Deciding Landing Page renders the login page when user is not logged in 1`] = `
<Redirect
  push={false}
  to="/login"
/>
`;

exports[`Deciding Landing Page renders the order list when user is logged in 1`] = `
<Redirect
  push={false}
  to="/my_orders"
/>
`;
